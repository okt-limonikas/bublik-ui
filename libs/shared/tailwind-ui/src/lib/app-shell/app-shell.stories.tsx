/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { StoryFn, Meta } from '@storybook/react';
import { AppShell } from './app-shell';

export default {
	component: AppShell,
	title: 'components/App Shell',
	parameters: { layout: 'fullscreen' }
} as Meta<typeof AppShell>;

const Template: StoryFn<typeof AppShell> = (args) => (
	<AppShell
		sidebar={
			<div className="w-[52px] bg-white h-screen flex flex-col items-center justify-center">
				<span
					style={{ textOrientation: 'upright', writingMode: 'vertical-rl' }}
				>
					Sidebar
				</span>
			</div>
		}
	>
		<div className="grid h-screen p-1 place-items-center">
			<h1>Main content</h1>
		</div>
	</AppShell>
);

export const Primary = {
	render: Template,
	args: {}
};
