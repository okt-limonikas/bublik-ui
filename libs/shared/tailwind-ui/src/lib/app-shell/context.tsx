/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { createContext, FC, ReactNode, useContext } from 'react';

export interface SidebarContext {
	isSidebarOpen?: boolean;
	toggleSidebar?: () => void;
}

export const SidebarContext = createContext<SidebarContext | null>(null);
SidebarContext.displayName = 'SidebarContext';

export const useSidebar = () => {
	const context = useContext(SidebarContext);

	if (!context) {
		throw new Error('useSidebar must be used within a SidebarContextProvider');
	}

	return context;
};

export interface SidebarProviderProps extends SidebarContext {
	children?: ReactNode;
}

export const SidebarProvider: FC<SidebarProviderProps> = ({
	isSidebarOpen = true,
	// eslint-disable-next-line @typescript-eslint/no-empty-function
	toggleSidebar = () => {},
	children
}) => {
	return (
		<SidebarContext.Provider value={{ isSidebarOpen, toggleSidebar }}>
			{children}
		</SidebarContext.Provider>
	);
};
