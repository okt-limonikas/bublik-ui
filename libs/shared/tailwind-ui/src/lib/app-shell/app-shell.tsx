/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { PropsWithChildren, ReactNode, useCallback } from 'react';
import { StoryFn } from '@storybook/react';

import { useLocalStorage, useKey } from '@/shared/hooks';
import { isFocusInInput } from '@/shared/utils';

import { SidebarProvider, useSidebar } from './context';
import { cn } from '../utils';

const SidebarStory = () => {
	const { isSidebarOpen } = useSidebar();

	return (
		<div
			className={cn(
				'sticky top-0 z-20 h-screen bg-white',
				isSidebarOpen ? 'w-[245px]' : 'w-[49px]'
			)}
		/>
	);
};

export const withSidebar = (hideSidebar?: boolean) => {
	return (Story: StoryFn) => (
		<AppShell sidebar={hideSidebar ? null : <SidebarStory />}>
			<Story />
		</AppShell>
	);
};

export interface AppShellProps {
	sidebar: ReactNode;
}

export const AppShell = ({
	sidebar,
	children
}: PropsWithChildren<AppShellProps>) => {
	const [isSidebarOpen, setIsSidebarOpen] = useLocalStorage(
		'sidebar-open',
		false
	);

	const toggleSidebar = useCallback(
		() => setIsSidebarOpen(!isSidebarOpen),
		[isSidebarOpen, setIsSidebarOpen]
	);

	useKey(
		(e) => e.code === 'KeyS' && !e.ctrlKey && !e.metaKey && !isFocusInInput(e),
		toggleSidebar,
		{ event: 'keypress' }
	);

	return (
		<div className="relative flex h-full" data-testid="tw-app-shell">
			<SidebarProvider
				isSidebarOpen={isSidebarOpen}
				toggleSidebar={toggleSidebar}
			>
				<div className="sticky top-0 z-20 h-screen h-svh" id="sidebar">
					{sidebar}
				</div>
				<div className="flex-grow overflow-auto" id="page-container">
					{children}
				</div>
			</SidebarProvider>
		</div>
	);
};
