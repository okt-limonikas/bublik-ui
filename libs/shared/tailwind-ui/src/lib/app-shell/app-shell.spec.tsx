/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { it, expect, describe } from 'vitest';
import { render } from '@testing-library/react';

import { AppShell } from './app-shell';

describe('components/AppShell', () => {
	it('should render successfully', () => {
		const { getByTestId } = render(<AppShell sidebar={<div>sidebar</div>} />);

		const badge = getByTestId('tw-app-shell');

		expect(badge).toBeVisible();
	});

	it('should match snapshot', () => {
		const { asFragment } = render(<AppShell sidebar={<div>sidebar</div>} />);

		expect(asFragment()).toMatchSnapshot();
	});
});
