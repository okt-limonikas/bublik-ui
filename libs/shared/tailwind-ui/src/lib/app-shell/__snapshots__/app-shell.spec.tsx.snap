// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/AppShell > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative flex h-full"
    data-testid="tw-app-shell"
  >
    <div
      class="sticky top-0 z-20 h-screen h-svh"
      id="sidebar"
    >
      <div>
        sidebar
      </div>
    </div>
    <div
      class="flex-grow overflow-auto"
      id="page-container"
    />
  </div>
</DocumentFragment>
`;
